File: backend-dotnet/Application/Commands/Posts/DeletePostCommandHandler.cs
using Application.Interfaces;
using Microsoft.Extensions.Logging;

namespace Application.Commands.Posts
{
    public class DeletePostCommandHandler : ICommandHandler<DeletePostCommand>
    {
        private readonly IPostRepository _postRepository;
        private readonly ICommentRepository _commentRepository;
        private readonly IPostLikeRepository _postLikeRepository;
        private readonly IPostTagRepository _postTagRepository;
        private readonly IBlobService _blobService;
        private readonly ICacheService _cacheService;
        private readonly ILogger<DeletePostCommandHandler> _logger;

        public DeletePostCommandHandler(
            IPostRepository postRepository,
            ICommentRepository commentRepository,
            IPostLikeRepository postLikeRepository,
            IPostTagRepository postTagRepository,
            IBlobService blobService,
            ICacheService cacheService,
            ILogger<DeletePostCommandHandler> logger)
        {
            _postRepository = postRepository;
            _commentRepository = commentRepository;
            _postLikeRepository = postLikeRepository;
            _postTagRepository = postTagRepository;
            _blobService = blobService;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<bool> HandleAsync(DeletePostCommand command)
        {
            _logger.LogInformation("Attempting to delete post {PostId} by user {UserId}", 
                command.PostId, command.UserId);

            var post = await _postRepository.GetPostWithDetailsAsync(command.PostId);

            if (post == null)
            {
                throw new KeyNotFoundException("Post not found");
            }

            if (post.UserId != command.UserId)
            {
                throw new UnauthorizedAccessException("You can only delete your own posts");
            }

            // Delete image from blob storage if exists
            if (!string.IsNullOrEmpty(post.ImageUrl))
            {
                try
                {
                    await _blobService.DeleteAsync(post.ImageUrl, "posts");
                    _logger.LogInformation("Image deleted from blob storage for post {PostId}", command.PostId);
                }
                catch (Exception ex)
                {
                    _logger.LogWarning(ex, "Failed to delete image from blob storage for post {PostId}", command.PostId);
                    // Continue with post deletion even if blob deletion fails
                }
            }

            // Remove related entities (EF will handle cascade delete, but let's be explicit)
            _postLikeRepository.RemoveRange(post.Likes);
            _commentRepository.RemoveRange(post.Comments);
            _postTagRepository.RemoveRange(post.PostTags);
            _postRepository.Remove(post);

            await _postRepository.SaveChangesAsync();

            // Clear caches
            await _cacheService.RemoveAsync($"post_details_{command.PostId}");
            _cacheService.RemoveByPattern("public_posts");
            _cacheService.RemoveByPattern($"user_feed_{command.UserId}");

            _logger.LogInformation("Post {PostId} deleted successfully", command.PostId);

            return true;
        }
    }
}

File: backend-java/src/main/java/com/jerrygram/application/commands/posts/DeletePostCommandHandler.java
package com.jerrygram.application.commands.posts;

import com.jerrygram.application.interfaces.IBlobService;
import com.jerrygram.application.interfaces.ICommandHandler;
import com.jerrygram.application.interfaces.IElasticService;
import com.jerrygram.infrastructure.repositories.PostRepository;
import com.jerrygram.infrastructure.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Slf4j
public class DeletePostCommandHandler implements ICommandHandler<DeletePostCommand, Boolean> {

    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final IBlobService blobService;
    private final IElasticService elasticService;

    @Override
    @Transactional
    public Boolean handle(DeletePostCommand command) {
        var postId = command.getPostId();
        var userId = command.getUserId();
        
        log.info("Deleting post: {} by user: {}", postId, userId);

        var post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));

        // Authorization check - only post owner can delete
        if (!post.getUser().getId().equals(userId)) {
            throw new SecurityException("User not authorized to delete this post");
        }

        // Delete image from blob storage if exists
        if (post.getImageUrl() != null && !post.getImageUrl().isEmpty()) {
            try {
                blobService.delete(post.getImageUrl());
                log.info("Image deleted from blob storage: {}", post.getImageUrl());
            } catch (Exception e) {
                log.warn("Failed to delete image from blob storage: {}", post.getImageUrl(), e);
            }
        }

        // Remove from Elasticsearch
        try {
            elasticService.deletePost(postId.toString());
            log.info("Post removed from Elasticsearch: {}", postId);
        } catch (Exception e) {
            log.warn("Failed to remove post from Elasticsearch: {}", postId, e);
        }

        // Delete the post (cascade will delete related entities, count will be automatically recalculated)
        postRepository.delete(post);
        
        log.info("Post {} deleted successfully by user: {}", postId, userId);
        return true;
    }
}